Add LPC11xx command to read out unique ID from target.

dlaw opened the pull request:
This pull request adds the "readuid" command, to read out the unique ID from an LPC11xx/LPC8xx target. I've tested it with a couple of LPC824s and verified that the UID printed in GDB matches the UID that I read out in my own code.

The function `lpc_iap_call()` did not support IAP commands with a return value, so I had to alter it slightly. I've attached the relevant page of the user manual for reference. The maximum number of `uint32`s returned by any IAP command is 4.

![iap_format](https://cloud.githubusercontent.com/assets/606010/24965611/907d6878-1f72-11e7-88c4-a8a6e5b1cd21.png)

dlaw commented:
Hi folks,

Could you let me know if you are interested in merging this PR?

I would be happy to make additional changes if you feel it is not suitable in its current state.

Cheers,
David

dlaw commented:
Dear Black Magic Probe maintainers,

Is there any timetable for evaluating this pull request? I remain available to make any changes or perform any additional tests that are requested. I just rebased my change onto the current head of master.

This functionality has become quite important to my work (I log the UID of every chip that is programmed by my Black Magic Probe setup).

Thank you for your consideration.

Sincerely,
David Lawrence

UweBonnes commented:
Gareth told about "September" for a review of the the patch and issue backlog...

gsmcmullin commented:
The code looks ok. I'm happy to merge.  The only problem with this is that making the IAP call may be problematic if the target is in an unknown state.  In the cases where these calls are made for programming flash, the target has just been reset, so we know what we're dealing with.  The monitor command could be given at any time.

I suggest at least save the memory used and restore after, and globally disable interrupts in the target while making the call.  Is there no way to read the unique ID my memory-mapped IO?

UweBonnes commented:
>  Is there no way to read the unique ID my memory-mapped IO?
Probably no documented way...

dlaw commented:
Thanks for the feedback, and for all your work on the Black Magic Probe. I will make the suggested modifications (to disable interrupts and restore the memory and registers which get clobbered). It may take me a few weeks to get it all polished and tested, because work is pretty busy right now.

I poked around and unfortunately the location of the unique ID in the boot ROM varies between different versions of the boot ROM, so the only safe way to read it is to use the API call. It is technically possible to get the UID using MMIO, by reading in the boot ROM and running it on the BMP in an emulator... but that seems like a lot more work than restoring state properly. :-)

dlaw commented:
After a *very* long delay (my apologies), we had a chance to revisit this pull request. Changes:
1. Rebased on master.
2. Added support for LPC15xx in addition to LPC8xx/LPC11xx.
3. Updated LPC IAP function call to save and restore registers and target RAM.
4. Tested on LPC824 and LPC1549.

Do you have any thoughts on the best way to disable interrupts during an IAP call? The DHCSR (debug halting control and status register) only permits interrupts to be disabled during single stepping. We could set FAULTMASK before the call and restore its value afterwards, but this still runs the risk of a non-maskable interrupt executing during an IAP call. And I am a little nervous about tinkering with so much of the target state and expecting it to be totally transparent to the target program.

Cheers,
David

UweBonnes commented:
Sorry, didn't notice the rebase and rework. It seem all issue that Gareth had are resolved now.
